src/Membrain/Units.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE CPP           #-}
+{-# LANGUAGE DataKinds     #-}
+#if ( __GLASGOW_HASKELL__ >= 806 )
+{-# LANGUAGE NoStarIsType  #-}
+#endif
+{-# LANGUAGE TypeOperators #-}
+
+{- | This module contains type aliases for memory data units.
+
+TODO: explain kilo/kibi difference.
+
+-}
+
+module Membrain.Units
+       ( -- * Simple
+         Bit
+       , Nibble
+       , Byte
+
+         -- * Decimal
+       , Kilobyte
+       , Megabyte
+       , Gigabyte
+       , Terabyte
+       , Petabyte
+       , Exabyte
+       , Zettabyte
+       , Yottabyte
+
+         -- * Binary
+       , Kibibyte
+       , Mebibyte
+       , Gibibyte
+       , Tebibyte
+       , Pebibyte
+       , Exbibyte
+       , Zebibyte
+       , Yobibyte
+
+         -- * Smart constructors
+       , bit
+       , nibble
+       , byte
+
+       , kilobyte
+       , megabyte
+       , gigabyte
+       , terabyte
+       , petabyte
+       , exabyte
+       , zettabyte
+       , yottabyte
+
+       , kibibyte
+       , mebibyte
+       , gibibyte
+       , tebibyte
+       , pebibyte
+       , exbibyte
+       , zebibyte
+       , yobibyte
+       ) where
+
+import GHC.Natural (Natural)
+import GHC.TypeNats (type (*))
+
+import Membrain.Memory (Memory, memory)
+
+
+type Bit       = 1
+type Nibble    = 4
+type Byte      = 8
+
+type Kilobyte  = 1000 * Byte
+type Megabyte  = 1000 * Kilobyte
+type Gigabyte  = 1000 * Megabyte
+type Terabyte  = 1000 * Gigabyte
+type Petabyte  = 1000 * Terabyte
+type Exabyte   = 1000 * Petabyte
+type Zettabyte = 1000 * Exabyte
+type Yottabyte = 1000 * Zettabyte
+
+type Kibibyte  = 1024 * Byte
+type Mebibyte  = 1024 * Kibibyte
+type Gibibyte  = 1024 * Mebibyte
+type Tebibyte  = 1024 * Gibibyte
+type Pebibyte  = 1024 * Tebibyte
+type Exbibyte  = 1024 * Pebibyte
+type Zebibyte  = 1024 * Exbibyte
+type Yobibyte  = 1024 * Zebibyte
+
+-- | Creates 'Memory' in 'Bit's from given 'Natural' as the number of bits.
+bit :: Natural -> Memory Bit
+bit = memory
```

Comment:
As a minor performance optimization, you can define `bit` as `Memory` constructor